File: Coding Blocks CP Course/Segment tree challenges/min_query.cpp
#include<bits/stdc++.h>
using namespace std;

class segmenttree
{
public:
    int N;
    
    segmenttree(int n){
        N=n;
    }
    void build(int l,int r,int node,vector<int> a){
        if(l==r){
            st[node]=a[l];
            return;
        }
        int mid=(l+r)/2;
        build(l,mid,node*2+1,a);
        build(mid+1,r,node*2+2,a);
        st[node]=min(st[2*node+1],st[2*node+2]);
    }

    void update(int l,int r,int indup,int val,int node){
        if(l==r){
            a[l]=val;
            st[node]=val;
            return;
        }
        else{
            int mid=(l+r)/2;
            if(indup>=l&&indup<=mid){
                update(l,mid,indup,val,node*2+1);
            }
            else{
                update(mid+1,r,indup,val,node*2+2);
            }
            st[node]=min(st[2*node+1],st[2*node+2]);  
        }
    }

    int query(int si,int se,int l,int r,int node){
        if(se<l||si>r||l>r){
            return INT_MAX;
        }
        if(si>=l&&se<=r){
            return st[node];
        }
        int mid=(si+se)/2;
        int q1=query(si,mid,l,r,node*2+1);
        int q2=query(mid+1,se,l,r,node*2+2);
        return min(q1,q2);
    }
};

int main(){
    int n,q;
    cin>>n>>q;
    segmentree T=segmenttree(n);
    vector<int> arr;
    int temp;
    for(int i=0;i<n;i++){
        cin>>temp;
        arr.push_back(temp);
    }

    T.build(0,n-1,0,arr);

    for(int i=0;i<4*n;i++){
        cout<<T.st[i]<<" ";
    }
    cout<<"\n";
    return 0;
}

File: Coding Blocks CP Course/Segment tree challenges/Min_Query.py
def buildTree(ST,arr,ss,se,nodeIndex,debug=True):
	if ss==se:
		ST[nodeIndex]=arr[ss]
		return
	mid=(ss+se)//2
	buildTree(ST,arr,ss,mid,2*nodeIndex+1,debug)
	buildTree(ST,arr,mid+1,se,2*nodeIndex+2,debug)
	ST[nodeIndex]=min(ST[2*nodeIndex+1],ST[2*nodeIndex+2])
	return

def query(ST,qs,qe,ss,se,nodeIndex,debug):
	if ss>qe or se<qs:
		return float('inf')
	if ss>=qs and se<=qe:
		return ST[nodeIndex]
	mid=(ss+se)//2
	l=query(ST,qs,qe,ss,mid,2*nodeIndex+1,debug)
	r=query(ST,qs,qe,mid+1,se,2*nodeIndex+2,debug)
	return min(l,r)

def update(ST,arrInd,updatedVal,ss,se,nodeIndex,debug):
	if ss>arrInd or se<arrInd:
		return
	if ss==se and ss==arrInd:
		ST[nodeIndex]=updatedVal
		return
	mid=(ss+se)//2
	update(ST,arrInd,updatedVal,ss,mid,2*nodeIndex+1,debug)
	update(ST,arrInd,updatedVal,mid+1,se,2*nodeIndex+2,debug)
	return

import sys
debug=False
n,q = map(int,input().split())
arr = list(map(int,input().split()))
ST = [sys.maxsize]*(4*n +1)
buildTree(ST,arr,0,n-1,0,debug)
if debug:
	print("ST",ST)
for i in range(q):
	Line=list(map(int,input().split()))
	if Line[0]==1:
		l,r=Line[1::]
		print(query(ST,l-1,r-1,0,n-1,0,debug))
	elif Line[0]==2:
		arrInd,updatedVal=Line[1::]
		update(ST,arrInd-1,updatedVal,0,n-1,0,debug)
		if debug:
			print("ST after update",ST)